Add zulu8

mistydemeo opened the pull request:
This imports a `zulu8` package. To start, I've just copied the existing upstream Homebrew cask package; if we need customizations, we can do whatever we want. After running `brew cask install homebrew/bootstrap/zulu8`, everything looks fine, and this Java version seems to be working properly.

MikeMcQuaid commented:
@mistydemeo Could this be submitted to https://github.com/Homebrew/homebrew-cask-versions?

mistydemeo commented:
@MikeMcQuaid It is already in that tap! I added a copy here in order to provide some control over version upgrades.

I've added a `brew bootstrap-jenv-java` command which installs casks from this tap, then configures `jenv` to look for them.

vroldanbet commented on cmd/brew-bootstrap-jenv-java:
```diff
@@ -0,0 +1,87 @@
+
+BASE_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+
+if ! which rbenv &>/dev/null; then
```
I believe this should be `which jenv`

mistydemeo replied:
Good catch, fixed!

vroldanbet commented on cmd/brew-bootstrap-jenv-java:
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
+# Installs Java.
```
what about `Installs Zulu JDK`?

mistydemeo replied:
Good idea, changed.

vroldanbet commented on cmd/brew-bootstrap-jenv-java:
```diff
@@ -0,0 +1,87 @@
+  abort 'Error: add `eval "$(jenv init -)"` to the end of your .bash_profile!'
+}
+
+abort_for_zsh() {
```
❤️

vroldanbet commented on cmd/brew-bootstrap-jenv-java:
```diff
@@ -0,0 +1,87 @@
+  if ! [[ -z "$JENV_VERSION" ]]; then
+    JAVA_REQUESTED="$JENV_VERSION"
+  else
+    JAVA_REQUESTED="$(jenv local)"
```
if `.java-version` does not exist, `jenv local` will fail. Should we check for this file and fail fast?

mistydemeo replied:
I don't think we'll get this far - in the case there's no `.java-version` and `$JENV_VERSION` is unset, `jenv version-name` will fall back to `system` and this entire `if` statement body will be skipped.

vroldanbet commented on cmd/brew-bootstrap-jenv-java:
```diff
@@ -0,0 +1,87 @@
+  abort "Error: you need to 'brew install jenv'!"
+fi
+
+if ! jenv version-name &>/dev/null; then
```
I'm not sure to understand the goal of this if clause 😕 

`jenv version-name` indicates the version of the currently configured JDK, following jEnv's priority semantics: if `.java-version` is present, then that's the version, otherwise fallback to the next in priority (No 💯 sure, it guess it would be the global JDK).

So if there is no JDK at all installed, `$JAVA_REQUESTED` won't be set, is that correct?
I think it would be OK to remove that wrapping if clause and just leave the ones inside, checking the env variable or a locally configured JDK

mistydemeo replied:
`jenv version-name` exits 1 if the requested version isn't known to jenv; this is tracking that status rather than checking the value. We have the same logic in the rbenv and nodenv versions of the script.

vroldanbet replied:
gotcha, so if it's known, wouldn't `$JAVA_REQUEST` not be defined, and if so, the script would fail in subsequent lines?

mistydemeo replied:
`$JAVA_REQUESTED` is only assigned and used within the body of the `if` statement, which won't be reached if `jenv version-name` exits 1.

vroldanbet replied:
duh, me dufus. Just realized the identation of the nested statements. Don't mind my comment 😓

vroldanbet commented on cmd/brew-bootstrap-jenv-java:
```diff
@@ -0,0 +1,87 @@
+cask "$cask_name"
+EOS
+
+  java_home=$(/usr/libexec/java_home -v $JAVA_REQUESTED)
```
totally didn't know about `java_home`, so useful! ❤️

vroldanbet commented on cmd/brew-bootstrap-jenv-java:
```diff
@@ -0,0 +1,87 @@
+EOS
+
+  java_home=$(/usr/libexec/java_home -v $JAVA_REQUESTED)
+  yes | jenv add $java_home
```
that piped `yes` would override any configured JDK with the same version, right?

mistydemeo replied:
That's right! The idea here is that, because `jenv add` exits 1 for versions/paths that it already tracks, this should overwrite the user's existing config. That covers the cases where

1. The user already had Zulu installed, but part of it wasn't known to jenv, or
2. The user already had a different JVM installed and tracked via jenv.

mistydemeo commented:
> Also: I'm wondering if this PR suggests that we are only providing initially the JDK 8 package? I imagine that new and recent Java projects would like to use Java 11, our new "LTS". Is this something to be tackled in a different PR (like, to add a zulu versioned package for each major release?)

I can add it in this PR. IIRC I opened the PR before Zulu's Java 11 package had been released, but it's out now! It isn't in Cask proper yet, but I can make a versioned package based on what we've got.

vroldanbet commented:
@mistydemeo awesome, thank you! I think we can do that in a follow up PR. Let's wait for Zulu to release 11. I think this is ready to be merged 👍
